stan/math/rev/functor/ode_adjoint.hpp

```diff
@@ -0,0 +1,172 @@
+ * passed to CVODES
+ * @param absolute_tolerance_forward Absolute tolerance per ODE state for
+ * forward problem passed to CVODES
+ * @param relative_tolerance_backward Relative tolerance for backward problem
+ * passed to CVODES
+ * @param absolute_tolerance_backward Absolute tolerance per ODE state for
+ * backward problem passed to CVODES
+ * @param relative_tolerance_quadrature Relative tolerance for quadrature
+ * problem passed to CVODES
+ * @param absolute_tolerance_quadrature Absolute tolerance for quadrature
+ * problem passed to CVODES
+ * @param max_num_steps Upper limit on the number of integration steps to
+ *   take between each output (error if exceeded)
+ * @param num_steps_between_checkpoints Number of integrator steps after which a
+ * checkpoint is stored for the backward pass
+ * @param interpolation_polynomial type of polynomial used for interpolation
+ * @param solver_forward solver used for forward pass
+ * @param solver_backward solver used for backward pass
+ * @param[in, out] msgs the print stream for warning messages
+ * @param args Extra arguments passed unmodified through to ODE right hand side
+ * @return An `std::vector` of Eigen column vectors with scalars equal to
+ *  the least upper bound of `T_y0`, `T_t0`, `T_ts`, and the lambda's arguments.
+ *  This represents the solution to ODE at times \p ts
+ */
+template <typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename T_abs_tol_fwd, typename T_abs_tol_bwd, typename... T_Args,
+          require_all_eigen_col_vector_t<T_y0, T_abs_tol_fwd,
+                                         T_abs_tol_bwd>* = nullptr>
+auto ode_adjoint_impl(
+    const char* function_name, F&& f, const T_y0& y0, const T_t0& t0,
+    const std::vector<T_ts>& ts, double relative_tolerance_forward,
+    const T_abs_tol_fwd& absolute_tolerance_forward,
+    double relative_tolerance_backward,
+    const T_abs_tol_bwd& absolute_tolerance_backward,
+    double relative_tolerance_quadrature, double absolute_tolerance_quadrature,
+    long int max_num_steps,                  // NOLINT(runtime/int)
+    long int num_steps_between_checkpoints,  // NOLINT(runtime/int)
+    int interpolation_polynomial, int solver_forward, int solver_backward,
+    std::ostream* msgs, const T_Args&... args) {
+  using integrator_vari
+      = cvodes_integrator_adjoint_vari<F, plain_type_t<T_y0>, T_t0, T_ts,
+                                       plain_type_t<T_Args>...>;
+  auto integrator = new integrator_vari(
+      function_name, std::forward<F>(f), y0, t0, ts, relative_tolerance_forward,
+      absolute_tolerance_forward, relative_tolerance_backward,
+      absolute_tolerance_backward, relative_tolerance_quadrature,
+      absolute_tolerance_quadrature, max_num_steps,
+      num_steps_between_checkpoints, interpolation_polynomial, solver_forward,
+      solver_backward, msgs, args...);
+  return integrator->solution();
+}
```

Comment:
Why does this impl version exist? aka could we just move the construction of the vari to `ode_adjoint_tol_ctl`?

Reply:
Yes we could get rid of ìmpl`. It's just that we definitely want to include more ode_adjoint signatures with less tuning parameters. This will happen once we know how to simplify them in a meaningful way.